models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+}
+
+nest::glif_psc::Buffers_::Buffers_( glif_psc& n )
+  : logger_( n )
+{
+}
+
+nest::glif_psc::Buffers_::Buffers_( const Buffers_&, glif_psc& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::glif_psc::glif_psc()
+  : Archiving_Node()
+  , P_()
+  , S_( P_ )
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::glif_psc::glif_psc( const glif_psc& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::init_state_( const Node& proto )
+{
+  const glif_psc& pr = downcast< glif_psc >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::glif_psc::init_buffers_()
+{
+  B_.spikes_.clear();   // includes resize
+  B_.currents_.clear(); // include resize
+  B_.logger_.reset();   // includes resize
+}
+
+void
+nest::glif_psc::calibrate()
+{
+  B_.logger_.init();
+
+  V_.t_ref_remaining_ = 0.0;
+  V_.t_ref_total_ = P_.t_ref_;
+  V_.last_spike_ = 0.0;
+  V_.last_voltage_ = 0.0;
+
+  // post synapse currents
+  const double h = Time::get_resolution().get_ms(); // in ms
+
+  V_.P11_.resize( P_.n_receptors_() );
+  V_.P21_.resize( P_.n_receptors_() );
+  V_.P22_.resize( P_.n_receptors_() );
+  V_.P31_.resize( P_.n_receptors_() );
+  V_.P32_.resize( P_.n_receptors_() );
+
+  S_.y1_.resize( P_.n_receptors_() );
+  S_.y2_.resize( P_.n_receptors_() );
+  V_.PSCInitialValues_.resize( P_.n_receptors_() );
+
+  B_.spikes_.resize( P_.n_receptors_() );
+
+  double Tau_ = P_.C_m_ / P_.G_; // in ms
+  V_.P33_ = std::exp( -h / Tau_ );
+  V_.P30_ = 1 / P_.C_m_ * ( 1 - V_.P33_ ) * Tau_;
+
+  for ( size_t i = 0; i < P_.n_receptors_(); i++ )
+  {
+    // these P are independent
+    V_.P11_[ i ] = V_.P22_[ i ] = std::exp( -h / P_.tau_syn_[ i ] );
+
+    V_.P21_[ i ] = h * V_.P11_[ i ];
+
+    // these are determined according to a numeric stability criterion
+    // input time parameter shall be in ms, capacity in pF
+    V_.P31_[ i ] = propagator_31( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+    V_.P32_[ i ] = propagator_32( P_.tau_syn_[ i ], Tau_, P_.C_m_, h );
+
+    V_.PSCInitialValues_[ i ] = 1.0 * numerics::e / P_.tau_syn_[ i ];
+    B_.spikes_[ i ].resize();
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::glif_psc::update( Time const& origin, const long from, const long to )
+{
+  // glif_lif_r_asc_a
+  const double dt = Time::get_resolution().get_ms();
+
+  // get model type
+  std::string model_str = P_.glif_model_;
+  std::transform( model_str.begin(), model_str.end(), model_str.begin(), ::tolower );
```

Comment:
We should not support carelessness on part of the user. The user should use exact spelling, so no automatic "tolower" or similar. But see my suggestion on handling this in a different way in a separate comment in the h-file.